src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java

```diff
@@ -329,9 +329,22 @@ public ResponseEntity<ProgrammingExerciseStudentParticipation> requestFeedback(@
         return ResponseEntity.ok().body(participation);
     }
 
-    private boolean isNotAllowedToStartProgrammingExercise(ProgrammingExercise programmingExercise, @Nullable StudentParticipation participation) {
-        return participation != null ? exerciseDateService.isAfterDueDate(participation)
+    /**
+     * Checks if the student is currently allowed to participate in the course exercise using this participation
+     * @param programmingExercise the exercise where the user wants to participate
+     * @param participation       the participation, may be null in case there is none
+     * @return a boolean indicating if the user may participate
+     */
+    private boolean isAllowedToParticipateInProgrammingExercise(ProgrammingExercise programmingExercise, @Nullable StudentParticipation participation) {
+        boolean isAfterDueDate = participation != null ? exerciseDateService.isAfterDueDate(participation)
                 : (programmingExercise.getDueDate() != null && now().isAfter(programmingExercise.getDueDate()));
+        if (participation != null) {
+            // Test runs may only be started after the due date
+            return participation.isTestRun() == isAfterDueDate;
+        }
+        else {
+            return !isAfterDueDate;
+        }
```

Comment:
```suggestion
        if (participation != null) {
            // only regular participation before the due date; only practice run afterwards
            return participation.isTestRun() == exerciseDateService.isAfterDueDate(participation);
        }
        else {
            return programmingExercise.getDueDate() == null || now().isBefore(programmingExercise.getDueDate());
        }
```
I think the `?:` can be avoided here as the check for the `if` is the same to make the code more readable.

Please double-check that the suggestion does actually implement the same logic.

Reply:
Your version looks better and works as well